pubsub/awspubsub/awspubsub.go

```diff
@@ -135,41 +136,25 @@ var stringDataType = aws.String("String")
 
 // SendBatch implements driver.Topic.SendBatch.
 func (t *topic) SendBatch(ctx context.Context, dms []*driver.Message) error {
-	// The result for each message (either nil or an error) will be written to
-	// errc and collected at the end. We'll return the first non-nil error
-	// (if any).
-	errc := make(chan error, len(dms))
+	g, ctx := errgroup.WithContext(ctx)
 	var ctxErr error
 	for _, dm := range dms {
-		// If ctx is done, don't start any more goroutines. We can't just return
-		// because we want to make sure we wait until any goroutines we've already
-		// started complete.
-		if ctxErr != nil {
-			errc <- ctxErr
-			continue
-		}
 		select {
 		case <-ctx.Done():
 			ctxErr = ctx.Err()
-			errc <- ctxErr
+			break
 		case t.sem <- struct{}{}:
-			go func(dm *driver.Message) {
-				errc <- t.sendMessage(ctx, dm)
-				<-t.sem
-			}(dm)
+			dm := dm // https://golang.org/doc/faq#closures_and_goroutines
```

Comment:
I actually prefer this to passing args to closures in all cases, so I'm glad you were forced into it here :)

I don't think you need the comment -- it's a standard Go technique -- but I'm fine with it.